src/lib/checkbox/checkbox-foundation.ts

```diff
@@ -1,209 +1,194 @@
-  private _handleAnimationEnd: () => void;
+  private get _submittedValue(): string | null {
+    return this._checked ? this._value : null;
+  }
+  private get _formState(): CheckboxState {
+    return this._checked
+      ? this._indeterminate
+        ? 'checked-indeterminate'
+        : 'checked'
+      : this._indeterminate
+        ? 'unchecked-indeterminate'
+        : 'unchecked';
+  }
 
-  constructor(private _adapter: ICheckboxAdapter) {}
+  // Listeners
+  private readonly _changeListener: EventListener;
+  private readonly _inputSlotListener: EventListener;
 
-  public disconnect(): void {
-    this._adapter.unlisten('change', this._handleChange);
-    this._adapter.disconnect();
-    this._adapter.uninstallPropertyChangeHooks();
-    this._adapter.unlisten('animationend', this._handleAnimationEnd);
+  constructor(private _adapter: ICheckboxAdapter) {
+    this._changeListener = (evt: Event) => this._handleChange(evt);
+    this._inputSlotListener = () => this._handleInputSlotChange();
   }
 
-  public connect(): void {
+  public initialize(): void {
     this._adapter.initialize();
-    this._currentCheckState = this._determineCheckState();
-    this._setDense(this.dense);
-    this._updateAriaChecked();
-    this._inputAttributeChangedListener = (name, value) => this._handleInputAttributeChange();
-    this._adapter.setInputAttributeObserver((name, value) => this._inputAttributeChangedListener(name, value));
-
-    this._handleChange = () => this._handleInputChange();
-    this._adapter.listen('change', this._handleChange);
-    this._adapter.installPropertyChangeHooks(() => {
-      this._handleInputChange();
-    });
-    this._handleAnimationEnd = () => this.handleAnimationEnd();
-    this._adapter.listen('animationend', this._handleAnimationEnd, true);
+    this._adapter.addRootListener('change', this._changeListener);
+    this._adapter.addInputSlotListener(this._inputSlotListener);
+    this._adapter.syncValue(this._submittedValue, this._formState);
+  };
 
-    this._handleInputChange();
+  public syncValidity(hasCustomValidityError: boolean): void {
+    this._adapter.syncValidity(hasCustomValidityError);
   }
 
-  public get dense(): boolean {
-    return this._dense;
-  }
-
-  public set dense(value: boolean) {
-    if (this._dense !== value) {
-      this._dense = value;
-      this._setDense(value);
-    }
+  public setValidity(flags?: ValidityStateFlags | undefined, message?: string | undefined): void {
+    this._adapter.setValidity(flags, message);
   }
 
-  public handleAnimationEnd(): void {
-    if (!this._enableAnimationEndHandler) {
+  private _handleChange(evt: Event): void {
+    if (this.readonly) {
+      this._adapter.setChecked(this._checked);
+      evt.stopPropagation();
       return;
     }
 
-    const handler: TimerHandler = () => {
-      this._adapter.removeRootClass(this._currentAnimationClass);
-      this._enableAnimationEndHandler = false;
-    };
+    this._adapter.redispatchEvent(evt);
```

Comment:
The `event.target.checked` property on this event doesn't appear to reflect the "new" state in the listener. We should be able to query the target element from the listener for its checked state and receive the new state.

Also, it appears that we cannot cancel the `input` or `change` event (since the native events do not allow that). It seems that we should be consistent with the switch here and allow for `preventDefault()`?

Reply:
I updated `redispatchEvent` to accept option overrides for situations like this